pkg/kubelet/dockershim/docker_container_cleanup_manager.go

```diff
@@ -0,0 +1,159 @@
+}
+
+func newDockerContainerCleanupManager(cleanupService containerCleanupService) *containerCleanupManager {
+	return &containerCleanupManager{
+		cleanupService: cleanupService,
+		cleanupInfos:   orderedmap.New(),
+		mutex:          &sync.Mutex{},
+	}
+}
+
+// start makes the manager periodically review the cleanup infos it knows about and clean up the
+// ones that become stale.
+// This is a blocking call.
+// It can be passed up to do channels: the first one will stop the manager when closed;
+// The second one will be sent to after each "tick", ie call to cleanupStaleContainerCleanupInfos,
+// giving the IDs of containers that got cleaned up, if any.
+// These channels are mainly intended for tests, and both can be left nil.
+func (cm *containerCleanupManager) start(stopChannel <-chan struct{}, tickChannel chan []string) {
+	if stopChannel == nil {
+		stopChannel = wait.NeverStop
+	}
+
+	wait.Until(func() {
+		containerIDs := cm.cleanupStaleContainerCleanupInfos()
+		if tickChannel != nil {
+			tickChannel <- containerIDs
+		}
+	}, staleContainerCleanupInterval, stopChannel)
+}
+
+// Having those as variables makes them easy to mock in unit tests.
+var (
+	currentNanoTimestampFunc = func() int64 {
+		return time.Now().UnixNano()
+	}
+
+	// the period in between 2 calls to cleanupStaleContainers once started
+	staleContainerCleanupInterval = 5 * time.Minute
+
+	// cleanup infos older than this much will be considered stale, and cleaned up
+	staleContainerCleanupInfoAge = 1 * time.Hour
+)
+
+// insert allows keeping track of a new container's cleanup info.
+func (cm *containerCleanupManager) insert(containerID string, cleanupInfo *containerCleanupInfo) {
+	if cleanupInfo == nil {
+		return
+	}
+
+	cm.mutex.Lock()
+	defer cm.mutex.Unlock()
+
+	if _, present := cm.cleanupInfos.Get(containerID); present {
+		// shouldn't happen, as we only insert at container creation time - but to err on the side of
+		// caution, let's delete from and re-insert into the ordered map to ensure that the order
+		// stays chronological
+		klog.Errorf("duplicate cleanup info for container ID %q", containerID)
+		cm.cleanupInfos.Delete(containerID)
+	}
+
+	cm.cleanupInfos.Set(containerID, &containerCleanupInfoWithTimestamp{
+		cleanupInfo: cleanupInfo,
+		timestamp:   currentNanoTimestampFunc(),
+	})
+}
+
+func (cm *containerCleanupManager) performCleanup(containerID string) {
+	cm.mutex.Lock()
+	defer cm.mutex.Unlock()
+
+	cm.unsafePerformCleanup(containerID)
+}
+
+// unsafePerformCleanup is the same as performCleanup, but assumes the manager's lock has already
+// been acquired.
+func (cm *containerCleanupManager) unsafePerformCleanup(containerID string) {
+	if cleanupInfoWithTimestamp, present := cm.cleanupInfos.Delete(containerID); present {
+		cleanupInfo := cleanupInfoWithTimestamp.(*containerCleanupInfoWithTimestamp).cleanupInfo
+		cm.cleanupService.performContainerCleanup(containerID, cleanupInfo)
```

Comment:
Should `performContainerCleanup` bubble up the error from `performPlatformSpecificContainerCleanup` here so we can check for errors and insert back `<containerID, cleanupInfo>` to `cleanupInfos` for retrying the cleanup in future?